### DISCUSSION TITLE
How to add a new prop to eject component in the TypeScript project

### DISCUSSION THREAD

**Discussion by jcubic** [facebook/docusaurus#9883] [Q&A] [answered]
When you swizzle you get JavaScript files. But the TypeScript definition is fixed so you can't add a new prop to the component when ejecting with `swizzle`. I want to add a theme prop to overwrite the default prism theme in the codeblock. I want a different theme for the Home page.

```javascript
const prismTheme = theme ?? usePrismTheme();
```

But when using:

```jsx
import { themes as prismThemes } from 'prism-react-renderer';

<CodeBlock language="scheme" theme={prismThemes.github}>
  {example.code}
</CodeBlock>
```

I got an error that there is not a valid prop.

Is there a way to update the typeScript definition of the component that was ejected?

I've tried this (found in some closed issue):

```bash
npm run swizzle @docusaurus/theme-classic --typescript
```

But it doesn't work and ejected files are JavaScript files.

**Comment by slorber** [comment]
please create a repro of the problem and I'll show you a possible solution if I have one

**Reply by jcubic** [reply]
Here is a repro: https://codesandbox.io/p/devbox/fragrant-resonance-y7n9vp?file=%2Fsrc%2Fpages%2Findex.tsx

Note that you can't add `--typescript` to the swizzle command. But anyway, the option is ignored.

**Reply by slorber** [reply]
![CleanShot 2024-02-24 at 18 53 12](https://github.com/facebook/docusaurus/assets/749374/6ee3c1e3-0451-41c4-82ec-76eabf72ae0b)

**Reply by jcubic** [reply]
It seems that every time you create a sandbox with docusaurs.new it's private.

Try now:
https://codesandbox.io/p/devbox/fragrant-resonance-y7n9vp?file=%2Fsrc%2Fpages%2Findex.tsx

**Reply by slorber** [reply, marked as answer]
To be honest I'm not a fan of the CodeSandbox UX lately, and only use StackBlitz now.

You can use TS interface declaration merging feature to "enhance" the existing theme props types and add yours. Normally all our theme components export a `Props` interface.

This compiles with:

```ts
declare module "@theme/CodeBlock" {
  export interface Props {
    readonly theme: PrismTheme;
  }
}
```

https://codesandbox.io/p/devbox/exciting-mountain-fmh3z6?file=%2Fsrc%2Fpages%2Findex.tsx%3A62%2C2&workspaceId=cfc1fcc6-71a5-4f69-aa36-2cb4bfd3697d

**Reply by jcubic** [reply]
Now your link gives 404. But it looks like this is what I need. Thanks.
